lablup/backend.ai
# feat: Impl dummy agent

resolves #1283 

- dummy agents spawn dummy kernels.
- dummy kernels does not affect to agent's resource stat.
- dummy agent's kernels create one of two kind of code runner, `DummyCodeRunner` or `DummyFakeCodeRunner`. `DummyCodeRunner` initiates ZMQ sockets and implements only abstract methods of `AbstractCodeRunner`. 
`DummyFakeCodeRunner` does not initiate ZMQ sockets and it overrides all methods of `AbstractCodeRunner`.
- we can mimic dummy agent's time-consuming steps by setting configuration file `dummy-agent.toml`. we can set the delay of each step by a tuple of 2 numbers that will convert to a range of random float or just a float that will convert to a delay time(sec).
- dummy agents do not run `sync_container_lifecycles`.

## How to test
1. make a config file to `<BACKENDAI_ROOT>/dummy-agent-config.toml` like below and set each value. You can find it from `<BACKENDAI_ROOT>/configs/agent/sample-dummy-config.toml`
```
[agent.delay]
scan-image = 0.1
pull-image = 0.1
destroy-kernel = 0.1
clean-kernel = 0.1
create-network = 0.1
destroy-network = 0.1


[agent.image]
already-have = []
need-to-pull = []
missing = []


[agent.resource.cpu]
core-indexes = [0, 1, 2, 3, 4]


[agent.resource.memory]
size = 34359738368


[kernel-creation-ctx.delay]
prepare-scratch = 0.1
prepare-ssh = 0.1
spawn = 0.1
start-container = 2.0
mount-krunner = 0.1


[kernel]
use-fake-code-runner = true


[kernel.delay]
check-status = 0.1
get-completions = 0.1
get-logs = 0.1
interrupt-kernel = 0.1
start-service = 0.1
start-model-service = 0.1
shutdown-service = 0.1
commit = 10.0
get-service-apps = 0.1
accept-file = 0.1
download-file = 0.1
download-single = 0.1
list-files = 0.1
```

2. Set backend config at `agent.toml` to `dummy`
```
[agent]
mode = "dummy" # or backend = "dummy"
rpc-listen-addr = { host = "127.0.0.1", port = 6011 }
...
```
